recipe-client-addon/lib/StudyStorage.jsm

```diff
@@ -0,0 +1,210 @@
+let gDatabase;
+
+const DB_OPTIONS = {
+  namespace: "shield-study-addons",
+  version: 1,
+  storage: "persistent",
+  schema: {
+    name: { type: "string", unique: true },
+    version: { type: "string" },
+    description: { type: "string" },
+    studyStartDate: { type: "number" },
+    studyEndDate: { type: "number" },
+    addonId: { type: "string", },
+    active: { type: "boolean" },
+  },
+};
+
+// Compares a given `data` object against the DB_OPTIONS.schema. Throws if a key
+// that is not in the schema appears, or if the schema `type` does not match the
+// given value type.
+function validateSchema(data) {
+  const {schema} = DB_OPTIONS;
+
+  Object.keys(data).forEach((dataKey) => {
+    const incomingData = data[dataKey];
+    const requiredType = schema[dataKey] && schema[dataKey].type;
+
+    if (!requiredType) {
+      throw new Error(`Key "${dataKey}" does not exist in Study schema.`);
+    }
+
+    // In some cases, the value will be empty, at which point we just want to
+    // skip the rest of validation.
+    if (typeof incomingData === "undefined" || incomingData === null) {
+      return;
+    }
+
+    // If the `type` is a function, it's going to be a class (like Map or Set),
+    // otherwise we can assume it is a primitive type (like "boolean" or "number").
+    const validates = typeof requiredType === "function" ?
+      incomingData instanceof requiredType : typeof incomingData === requiredType;
+
+    if (!validates) {
+      throw new Error(`Key "${dataKey}" must be of type "${requiredType.name || requiredType}"`);
+    }
+  });
+}
+
+
+// Creates new object store and populates that store's indexes.
+function setupDatabase(db) {
+  const {schema} = DB_OPTIONS;
+  const store = db.createObjectStore(DB_NAME, { keyPath: "name" });
+
+  Object.keys(schema).forEach((schemaKey) => {
+    store.createIndex(schemaKey, schemaKey, schema[schemaKey]);
+  });
+}
+
+// Determines if a DB connection/request already exists.
+function hasDatabaseRequest() {
+  return !!gDatabase;
+}
+
+// Returns the currently active IDB connection, and creates one if not
+// already present.
+async function getDatabase() {
+  if (!hasDatabaseRequest()) {
+    // gDB is a Promise to prevent multiple calls to `IDB.open` - subsequent calls
+    // will refer to this same promise, requiring only one `open` call internally.
+    gDatabase = new Promise(async (resolve) => {
+      const db = await IndexedDB.open(DB_NAME, DB_OPTIONS, setupDatabase);
+      resolve(db);
+    });
+  }
+
+  return gDatabase;
+}
+
+
+this.StudyStorage = {
+  /**
+   * Clears all data from the connected IndexedDB.
+   * @async
+   */
+  async clearAllData() {
+    const db = await getDatabase();
+    await db.objectStore(DB_NAME, "readwrite").clear();
+  },
+
+  /**
+   * If a database exists and is connected, disconnect and remove the reference
+   * to that database.
+   *
+   * @async
+   */
+  async closeConnection() {
+    if (hasDatabaseRequest()) {
```

Comment:
Thinking about it, this races if called multiple times, too. :-)

To prevent this, store `gDatabase` in a local (can just do this directly rather than calling `getDatabase()`, given we're directly nulling it out anyway), then null out the global, then wait for the local `db` ref, then await `db.close()`. This results in the global being nulled out synchronously from the first call, meaning any subsequent calls won't try to close it again.